lib/frp/src/core/node/wrapper.rs

```diff
@@ -0,0 +1,170 @@
+//! This module defines a common template structure used to define FRP nodes.
+
+use crate::prelude::*;
+
+use crate::data::*;
+use crate::debug::*;
+use crate::node::class::*;
+use crate::node::id::*;
+use crate::node::label::*;
+
+
+
+// ===================
+// === NodeWrapper ===
+// ===================
+
+// === NodeWrapper ===
+
+/// `NodeWrapper` is an outer layer for every FRP node. For example, the `Source<Out>` node is just
+/// an alias to `NodeWrapper<SourceShape<Out>>`, where `SourceShape` is it's internal representation.
+/// This struct bundles each node with information about target edges. Although the edges are used
+/// only to send events, they are bundled to every node type in order to keep the implementation
+/// simple.
+pub type NodeWrapper<Shape> = NodeWrapperTemplate<Shape,Output<Shape>>;
+
+impl<Shape:KnownOutput> NodeWrapper<Shape> {
+    /// Constructor.
+    pub fn construct<Label>(label:Label, shape:Shape) -> Self
+        where Label : Into<CowString> {
+        let data = NodeWrapperTemplateData::construct(label,shape);
+        let rc   = Rc::new(RefCell::new(data));
+        let this = Self {rc};
+        this.set_display_id(this.id());
+        this
+    }
+}
+
+impl<Shape,Out> NodeWrapperTemplate<Shape,Out> {
+    /// Sends an event to all the children.
+    pub fn emit_event(&self, event:&Out) {
+        self.rc.borrow().targets.iter().for_each(|target| {
+            target.on_event(event)
+        })
+    }
+}
+
+impl<Shape,T:Value>
+HasEventTargets for NodeWrapperTemplate<Shape,EventData<T>> {
+    fn add_event_target(&self, target:AnyEventConsumer<EventData<T>>) {
+        self.rc.borrow_mut().targets.push(target);
+    }
+}
+
+
+// === NodeWrapperTemplate ===
+
+/// Internal representation for `NodeWrapper`.
+#[derive(Debug,Derivative)]
+#[derivative(Default(bound="Shape:Default"))]
+#[derivative(Clone(bound=""))]
+#[allow(missing_docs)]
+pub struct NodeWrapperTemplate<Shape,Out> {
```

Comment:
Because NodeWrapper is defined as:
```rust
pub type NodeWrapper<Shape> = NodeWrapperTemplate<Shape,Output<Shape>>;
```

If we define it without this structure then `Output<Shape>` would need to be used as a field, and thus we would need to use bounds on a struct definition, which would make these bounds appear in much more places than now.

Reply:
I added docs aobut it.